lib/npo.src.js

```diff
@@ -77,6 +124,15 @@
 			cycle = timer(scheduling_queue.drain);
 		}
 	}
+	
+	function hasCatch(chain) {
+		for (var i = 0; i < chain.length; i++) {
+			if (chain[i].failure) {
+				return true;
+			}
+		}
+		return false;
+	}
```

Comment:
Just occurred to me, this logic may not be correct... but I'm not positive.

`chain` here is, perhaps, not well named in that it implies the entire chain of promises hanging off the current promise (think vertically), when in fact it's just a list of each of the individual promises that were chained off the current promise (think horizontally).

The reason this matters is, doing a loop only through the chain only asks if any of the calls against this promise registered a `catch` handler for it... but that's not the same thing as determining if the exception is "unhandled" or not... in fact, to determine if an exception is "unhandled", you have to let it try to propagate its way down (vertically) the chain, and only once it gets to the last promise, if it hasn't been caught, then it's "unhandled".

So I think maybe `hasCatch` needs just to be replaced with a check like `chain.length > 0`. the "notification" process already takes care of handing off a rejection to the chained promise(s), so that process would happen "recursively" until a rejection is either handled, or until `chain.length` was `0` (meaning we're at the end of the chain).

Does that make sense?